### ISSUE TITLE
Nominal Jobs

### ISSUE THREAD

**Issue by erictune** [kubernetes/kubernetes#14188]
Allow a Job to optionally be a  _Nominal_ Jobs by setting `.spec.nominal = true`.

If `nominal` is true, when each Pod is created, manager decides which "completion index" that pod is going to work on.  Manager ensures that there is one completion for each index from 0 to `completions - 1`.  We make it possible to expose that index via the downward API.  Pods can specialize themselves based on their index.  For example, by indexing into bash-array env var. 

Examples of podTemplates in _Nominal Jobs_:
```
spec:
  containers:
    - name: foo
      env:
        - name: N
          valueFrom:
            fieldRef:
              fieldPath: metadata.completionIndex
      cmd: [ "/bin/sh", "-c", "myprogram --work_unit=$N" ]
```
or
```
spec:
  containers:
    - name: foo
      env:
        - name: N
          valueFrom:
            fieldRef:
              fieldPath: metadata.completionIndex
      cmd: [ "/bin/sh", "-c",
             "files_to_process=(foo bar); myprogram --work_unit=$files_to_process[$N]" ]
```

**Comment by erictune** [comment]
Things to consider:

**Do we promise to never run two pods which are working on the same completion number at the same time?**
I don't think so.  In the case of network partitions, and the like, we would have to wait for the partitioned
pods to be confirmed dead before starting new pods to work on the same completion number.  That would hurt latency a lot in failures cases, I think.  This means that users should not rely on all pods at one moment in time having unique completion numbers.  Pods need to be idempotent or else ignore their completion number and lease work from a workqueue.

**Comment by mikedanese** [comment]
@soltysh

**Comment by pmorie** [comment]
@kubernetes/rh-cluster-infra

**Comment by mikedanese** [comment]
> "completion index"

To me this name means the order that task pods completed.

**Comment by bgrant0607** [comment]
Completion index shouldn't be in metadata, since it's job-specific.

`indexed` might be clearer than `nominal`

A predictable, dense index can be used to index into an array. UUIDs are sparse and unpredictable. They can be used for consistent hashing. 

Neither is needed for applications pulling from dynamic work allocation mechanisms, like task queues. This is for static work distribution.

**Comment by thockin** [comment]
We have thus far resisted doing this for run-forever jobs, asserting that sharding and indexing are better served as a helper component.  are we going back on that philosophy for jobs?  Does this preclude a job being resized (e.g. start with 10 replicas, then change that to 20 and somehow things figure it out)?

**Comment by soltysh** [comment]
@thockin why in your opinion resizing is a problem here? The author of the code running on a pod must be prepared to resize and handle bigger load/smaller load. If he don't that's a user error and not a problem k8s should handle. You can't prevent that from happening. Additionally I'm probably missing the rest of the context you're mentioning here and not doing the indexing/sharding, could you point me to previous discussion?
I also agree with @bgrant0607 the name `nominal` isn't too good here. I was scratching my head the first time I read it, what it is about. I'd rather prefer indexed, and have the downward API have the index-like sounding field you can access to get that info.

**Comment by bgrant0607** [comment]
@thockin If work distribution were static and one increased the number of completions, they'd have to increase the number of work items, however they represented that, yes. It could be a map in a ConfigData resource or just files in NFS.

I don't think all jobs would need this, but it may be convenient for some, as opposed to bringing up, say, celery + rabbitMQ.

Note that it's not tied to the resource name.

**Comment by thockin** [comment]
Specifically we avoided doing indexes for RC-owned pods because a) multiple
RCs could be driving a set of pods and b) because that encourages people to
assume things like [0] is special, which RCs don't want.

It may be that Job (sorry I have not followed closely enough) doesn't have
those problems.  If you grow a job, you'll have to make sure there are more
shards available in your sharded work.  What happens if you shrink a job?
Will it always kill pods with the highest index?

Will people abuse this to run indexed run-forever services?  Is that OK?

**Comment by bgrant0607** [comment]
Multiple RCs driving a set of pods: Do you mean multiple release tracks / canaries, or rolling update, or orphaning and adoption, or accidental overlap, or something else? I assume multiple release tracks under the same service, or rolling update. 

This is for batch jobs, which is a different scenario. Since they are relatively short in duration and complete specific work items, there's typically no need for multiple release tracks, and rolling update won't be implemented for jobs. 

There is a need for multiple levels of sharding, but that's compatible with this approach since each set of tasks will have different inputs and therefore a non-unified index space is just fine.

I can't imagine needing to hand over a subset of tasks from one controller to another. Handing over all of the tasks should be fine, because information about what tasks have completed is discoverable from the underlying pods.

Since job won't run pods with restartPolicy Always, it is unlikely that this could be abused for continuously running services.

The indices aren't tied to any resource names (pods or jobs), so they are freely reassignable in the case of failure or deletion/re-creation.

Yes, shrinking would have to kill off the greatest indices.

**Comment by davidopp** [comment]
> Will people abuse this to run indexed run-forever services?  Is that OK?

IMO if that happens, it would be a good sign that we should consider creating a run-forever object/controller that provides dense indexes.

> rolling update won't be implemented for jobs.

I understand the argument that this probably isn't necessary due to being short-lived, but is there any fundamental reason why we couldn't provide this? Unless I'm misunderstanding, all you need is a way to "hand off" ownership of an (already-running or not-yet-started) index from the old job controller to the new one. Or even simpler, you could build the capability into the job controller itself so you only need one job controller (similar to how deployment handles rolling update itself, but without there being another underlying controller(s) as in deployment).

**Comment by erictune** [comment]
Renamed to "Indexed Job" due to popular demand.

**Comment by erictune** [comment]
I've been thinking about how you would use Job in conjunction with a Task Queue, and I think it is actually pretty tricky to get right.   This definitely means we need documentation for how to do the Task Queue case, and it may mean that there is more of a need for the Static case, which is clearly easier for users to use.

We need a concrete example of how to write a parallel job that uses a work-queue.  I filed this issue #15982

**Comment by erictune** [comment]
I think we should do what this issue proposes for v1.

But, if we decide **not** to do it for v1, then I think the Job API can be simplified.

In particular, we could remove `.spec.completions`, because it is not needed.  Here is why:

- single pod case: only one completion needed for Job success.
- multiple pod case with task queue: as described in  #15982, all pods need to wait and not exit with success until their task queue is completely empty.  As soon as one pod signals this condition, none of the other pods can have outstanding leases, so they should be exiting soon and at any rate are safe to terminate.

**Comment by boulos** [comment]
First, thanks for renaming Nominal to Indexed.

I agree that ```.spec.completions``` is bizarre and seemingly unnecessary.  You either want to run to completion once (per item), or basically run until something is drained (which you can signal out of band or simply exit).  I'm not sure I agree about your "as soon as one pod signals", particularly in systems where work items can spawn new work: it's totally reasonable to ask for (up to) 200 workers, have them all use timeouts to determine "Okay, I guess no work is available, I'll exit" and have the true "Everyone is done" be in response to being the last person standing (i.e. no jobs are currently running to generate work).

Is this the right place to ask: where's the "I need 500 and if I can't get them all at once, fail"?  **Most** MPI-like things run this way, and all the commentary I'm seeing about expecting things to come-and-go or be idempotent suggests this isn't in plan...

**Comment by davidopp** [comment]
@boulos Gang scheduling was requested in #16845. Modulo that, I think this issue actually does what you want, with the possible exceptions that you want to specify the exact number of pods the system should create initially (which is also be the number of completions) rather than allowing it to choose the parallelism. Of course it cannot guarantee there will never be two of the same index running at the same time, as @erictune observed https://github.com/kubernetes/kubernetes/issues/14188#issuecomment-141507249
but we can design the feature so that in the absence of failures it does what you're describing.

**Comment by boulos** [comment]
@davidopp yeah, I didn't see that one linked anywhere (thanks).  And yes, as @erictune said earlier: if I'm running a Job for an hour or so, I'm not super concerned about partitions causing me to get into some weird state.  In particular, any system that relies on cooperation is going to realize his buddies are gone (e.g. Hadoop or MPI) and likely kill the whole thing.  We obviously don't promise it (your computer might catch fire), but I think it's still useful to describe what the promise of the system is in the absence of these failures.

So @erictune I think we **should** promise that we won't start more than 1 copy of your completion number (element? index?) unless a Pod / Container / Node "fails".  Then it should be a policy decision: how many retries (per index? per Job? probably both) and how soon do we consider failed (N.B. the default in Hadoop is insane at 10m) to avoid at least some class of intermittent failure.

**Comment by erictune** [comment]
I've written up examples of several different ways to use Job, not including Indexed Job, in #17345 (feedback on that PR is welcome).

Now that I have the additional experience working with Job, I think we should put Indexed Job on the shelf for now:
- No customer has asked for it.  
  - I originally argued for it, but now I am not sure it is needed. 
  - @smarterclayton tells me RedHat only uses it for serial use cases.
- Even after using it for several days, I found it confusing that there were so many different ways to use Job, and writing clear documentation was hard.
- I believe we can add it to the API later in a backward compatible way if we change our minds.
- I'm finding that other styles of using Job cover enough of the problem space that we don't need yet another solution.
  - For small numbers of parallel pods, like up to 20, creating multiple job single-completion objects from a template turns out to be really easy, and labels make it easy to track the jobs and the pods.
  - While researching work-queue-based solutions, it appears that many uses of work-queues assume
    continuous processing situations, not finite-duration situations.  A queue-length-driven autoscaler controlling an RC seems like a better abstraction than Job for those situations.
  - For more complex, finite duration, batch processing, people are probably using a framework like hadoop or spark.  For that, you can start one pod (via Job) that then launches the framework. 

@bgrant0607 @davidopp @soltysh any objections?

Once this is decided, we can think about what to do with `.spec.completions`.

**Comment by soltysh** [comment]
Totally agree with leaving Indexed Job behind until more clear use case appears.

> Once this is decided, we can think about what to do with .spec.completions.

What are you planning to do with `.spec.completions`?

**Comment by davidopp** [comment]
As long as we can add it later, I don't have any objection to dropping it for now. It's great that you did the research to understand how people are actually running batch jobs in practice.

> For small numbers of parallel pods, like up to 20, creating multiple job single-completion objects from a template turns out to be really easy, and labels make it easy to track the jobs and the pods.

What templating system did you use?

**Comment by bgrant0607** [comment]
If we drop completions, we should also drop parallelism. If we're only going to drop one, we should drop parallelism and keep completions.

**Comment by bgrant0607** [comment]
Use cases I've seen do things like generate a number of input files into a shared filesystem and then statically assign them to different workers by index. The files themselves are either indexed, or a map is put somewhere, such as in a config file. One could generate different arguments for each worker, but it's not necessary. Unless the number of workers is incredibly huge, passing the whole map to each worker isn't a big deal. A script can just select the line corresponding to its worker with something like `head -N map-file | tail -1`, for example.

**Comment by davidopp** [comment]
> One could generate different arguments for each worker, but it's not necessary. 
> A script can just select the line corresponding to its worker with something like head -N map-file | tail -1, for example.

Isn't the "N" a "different argument for each worker" ?

**Comment by davidopp** [comment]
I didn't understand why we are considering dropping parallelism and/or completions. That model (which assumes a work queue etc.) seems orthogonal to the one we are discussing here (@bgrant0607 's last comment described it). If we dropped parallelism and completions, what would we replace them with? Just a "replicas" like ReplicationController has? And only allow RestartAlways pods? In that case why have a Job abstraction at all?

**Comment by bgrant0607** [comment]
@davidopp I was responding to https://github.com/kubernetes/kubernetes/issues/14188#issuecomment-149728423

It's true that parallel jobs aren't needed/useful for dynamic work allocation scenarios, but I still think they're valid for static work allocation, which was what they were intended for when this API was proposed back in July.

**Comment by soltysh** [comment]
I'm still waiting for @erictune to explain his reasoning in removing `.spec.completions`. I do see a value in dynamic work allocation, which originates this issue, but instead of removing completions and parallelism, we should rather drop the defaults, so it's possible to leave those empty, just thinking out loud. Yet still the examples Eric wrote prove our current solution is a good fit for most of the use-cases. I'd be hesitant to modify our current implementation without a real demand, and I haven't seen one yet.

**Comment by bgrant0607** [comment]
Multiple completions were designed for static work allocation.

Use a ReplicationController for dynamic work allocation.

**Comment by bgrant0607** [comment]
See also #18016

**Comment by erictune** [comment]
Proposal is in #18260

**Comment by kargakis** [comment]
@kubernetes/sig-apps-feature-requests

**Comment by luckyfengyong** [comment]
@erictune Thanks for the initial work, just wondering do you still work on it?
@kargakis Any comment on the requirement?

**Comment by kent-at-multiscale** [comment]
Has there been any movement on this in two years?

I have a few use cases where I have a batch of jobs I need to execute. The size of the batch is static. Right now I have to do unnecessary scripting in order to generate N copies of a Job from a template with different names and have each one act on one index into a list of inputs. Having a mechanism where I could create one Job (template?) and create N copies of it where N is the size of the input array would drastically simplify my workflow.

In short, I have identical Jobs except for their names and one input parameter, which is a generated integer in the range [0, N). Why can't Kubernetes Jobs inject this ordinal value for me? The documentation teases us with the possibility, but then denies us with "**not implemented yet**". [https://kubernetes.io/docs/concepts/workloads/controllers/jobs-run-to-completion/#parallel-jobs](url)

**Comment by kent-at-multiscale** [comment]
If you want an idea for how to make this more appealing to users, you could provide a mechanism where a user provides a Job template and an array of input elements and have the Kubernetes controller generate one Job for each input element. For static work queues, having the input queue directly in the Job YAML or JSON (or perhaps as a ConfigMap) would be very easy to use and would satisfy a lot of use-cases. I know that would have saved me weeks of effort.

**Comment by k82cn** [comment]
I'm thinking if we can support Job related requirement in [kube-arbitrator](https://github.com/kubernetes-incubator/kube-arbitrator) which is target for batch related jobs.

**Comment by kent-at-multiscale** [comment]
I don't fully understand what `kube-arbitrator` is. The readme starts with "As a cluster admin", but what I am discussing is for a cluster user, not an admin. I have jobs to run, somebody else manages the cluster.

Could you elaborate a bit on what `kube-arbitrator` is?

**Comment by OPhamster** [comment]
similar use case to @kent-at-multiscale where I have cronJobs that need to be executed - but since the size of the batches are so large and rarely increase - it would be quite favourable to scale them horizontally by using a `job-index` - where the job-index can be used as an argument to pass to the job and hence work on that batch of data is performed. Having a work queue in this case with redis may be overkill with the added uncertainty whether the controller and batch jobs are scheduled and running at the right time(auto-scale effects).

**Comment by yaakov2** [comment]
I would like to support the proposal to inject the job id into the started containers.
In the discussion above, there is a statement that this feature "is not needed by actual customers". -- Let me explain why I need it:

My use case is that I need to process tens of millions of data points in a simple computation with a long pre-computation before processing the actual data points.  I plan to use a few thousand docker images to complete the computation in about ten minutes.

This case is quite trivial to implement with the proposed feature:  I divide the work into the right size of batches, start the dockers and each docker processes one large batch and writes the output into a shared data base.  In case of pod failure-and-restart, the process can find its batch and continue where it left off.

Without injecting the id into each image, the newly started image cannot know what job failed -- what work needs to be continued.  This makes this simple solution impossible.

I looked into using Google PubSub as a queue and found that PubSub does not reliably signal a drained queue: It is designed for continuous work and hence waits for new job.

In the end, the two "simple" approaches ALMOST work --- but fail just because of a "small feature" that is not supported:  kubernetes jobs does all needed --- except telling the pods their id number.  PubSub does all needed --- except telling the pods when they can shutdown.

As a result, I have to invest into a more complex solution.  That's a shame.  Given the lack of options, I may end up actually using a more complex solution --- adding another data point to the argument "People do not use this feature --- they use other solutions".  But this is only because the option of just using kubernetes Jobs lacks the last needed feature.

**Comment by soltysh** [comment]
@yaakov2 the reason we don't have this feature is because nobody had the time to write it, yet If you feel like both the community and the core k8s might benefit from it, you're more than welcome to open a PR implementing this functionality. 
Personally, I'd happy to support the feature and I can guide you with implementation.

**Comment by oleg-kiselev** [comment]
I have a situation nearly identical to [this issue](https://github.com/kubernetes/community/issues/458), which was closed in favor of "Indexed Job".  What is the status of this feature?   Currently, I am not finding anything in the k8s documentation that would allow the equivalent of having "restartPolicy: "OnFailure" _and_ distinct, stateful pod identities provided by either the existing StatefulSet or the proposed IndexedJob.  I am looking for a solution that works completely within the k8s framework, without adding external monitoring agents.

**Comment by yaakov2** [comment]
@soltysh -- Thanks for your offer to contribute to kubernetes.  I ended up writing my own job queue with the required feature: Properly signal the end of the work.

**Comment by ysaakpr** [comment]
What is the status of this issue? Is there way to get Pod Index in the parallel job with a `completion ` set. I am currently on Kubernetes 1.14.  Have to create parallel jobs with index to do associated tasks
